### ISSUE TITLE
Points calculation for ncs season

### ISSUE THREAD

**Issue by Joe2k** [india-ultimate/hub#154]
[Doc for calculation](https://docs.google.com/document/d/1Slb9s9InT35puuLM2xi4-UEbR0TRNffBaVF35oP2I7A/edit#heading=h.xoclil8gktb3)

**Comment by Joe2k** [comment]
Context: Competitions requested me to see if the point calculation can be integrated to the hub as currently they are doing it manually and it is a very tough and time taking process and only a few of them know to do it properly in the sheets

Have gone through this doc and feel like it will become super complex to do in the Hub as we have to keep track of many teams, if they are new or not and see old points. 

So basically would need a lot of linking and edge cases. 

Anyway putting it here to see if anyone else has any thoughts on this

**Comment by AnishKacham** [comment]
Oh yesss. It would be complex but super good.
I think we should start by adding NCS Season entity (table) in the DB that would capture the following relationships :
1. If a team will be participating in a particular NCS Season or not.
2. The tournaments that are part of an NCS Season.
3. The rosters of each team for a particular NCS Season (which would freeze at the start of the season?)

`3.` can be done later also as it isn't really needed for points calculation.

if we do `1.` and `2.` we could add a `SeasonStanding` entity that would capture points and standings of different NCS Teams,
 which would update at the end of every tournament of a season. We can use the tournament standing for the season standing updates. And finally when we click on an `End Current Season` button it would run a task that would output the final points.
 
 
Similarly when we are creating a new season, we can add `Use Previous Season Standings` `|` `Create a fresh Season Standing` 
 
 
We already have `Team` and `Tournament` entities. 

Finally we can add a ui for NCS separately on the hub? (Maybe initial version can just start by showing teams that are registered  for current season later extending it to show frozen roster also and  different regions each team played in etc, etc)

**Comment by punchagan** [comment]
> Context: Competitions requested me to see if the point calculation can be integrated to the hub as currently they are doing it manually and it is a very tough and time taking process and only a few of them know to do it properly in the sheets

A lot of competitions work seems to be like that. :joy: 

I think we shouldn't try to build this in for this season.  We should focus on smoothing out the workflow for memberships and running individual tournaments as much as we can, during this season. 

Competitions has history of data for NCS points from the time it has started. We should take some time to go through those, and the document to understand how to build-in the scoring system, and work on it for the next season.  The rules also keep changing between every year, slightly. Importing old data would help us how we can make the system flexible enough to accommodate for some kinds of changes, while making it easy to calculate the scores correctly. 

It would definitely be worth building, and displaying this data on the Hub. Having the historical data from old NCS seasons would be the cherry on the cake!

> I think we should start by adding NCS Season entity (table) in the DB that would capture the following relationships :
> 
>     1. If a team will be participating in a particular NCS Season or not.

This information would ideally come from teams rostering under the NCS event in UC (presently). But, ops/competitions don't enforce this. Also, different tournaments at different stages allow teams to directly participate in them, even if teams have not played previous tournaments in the season. But, any team that is rostered for any tournament of the season should be considered as participating. 

>     2. The tournaments that are part of an NCS Season.

This is new information that is not present on Ultimate Central - it is only in competitions committee's head or messages put out on Social Media. We definitely need to store this. 

>     3. The rosters of each team for a particular NCS Season (which would freeze at the start of the season?)

No, they are not frozen. People can be added removed during the season, etc. etc.

> We already have `Team` and `Tournament` entities.

We'd need a Season entity, which would also link to previous season to be able to get points for teams, etc.  There may be rules on what happens if teams merge/split up, etc., for the linking...

NCS has both mixed and single gender tournaments and points from the two single gender teams are added to the team's total. Also, clubs can have multiple teams between which players can be moved, etc. We need to see how the points for these teams get added, etc., to figure out if the data model needs a Club entity as well.  

The mapping of single gender (SG) tournaments from UC to the Hub also would be something we need to think about. This may be something we need to figure out, even if we don't work on this issue immediately. We need to look at how SG events and teams were created in UC and see if we need to make any changes on the Hub's end to be able to map those events and tournaments into the Hub.

**Comment by Joe2k** [comment]
Yep correct there are lot of cases and rules around them. We should think about them later properly.
